sklearn/tests/metadata_routing_common.py

```diff
@@ -25,55 +26,69 @@
 from sklearn.utils.multiclass import _check_partial_fit_first_call
 
 
-def record_metadata(obj, method, record_default=True, **kwargs):
-    """Utility function to store passed metadata to a method.
+def record_metadata(obj, record_default=True, **kwargs):
+    """Utility function to store passed metadata to a method of obj.
 
     If record_default is False, kwargs whose values are "default" are skipped.
     This is so that checks on keyword arguments whose default was not changed
     are skipped.
 
     """
+    stack = inspect.stack()
+    method = stack[1].function
+    parent = stack[2].function
     if not hasattr(obj, "_records"):
         obj._records = {}
+    if method not in obj._records:
+        obj._records[method] = {}
+    if parent not in obj._records[method]:
+        obj._records[method][parent] = []
```

Comment:
I think that we can simplified what is written here. I really like the `collections.defaultdict` for this initialization usage:

```suggestion
        obj._records = defaultdict(lambda: defaultdict(list))
```